pkg/resources/user_grant.go

```diff
@@ -0,0 +1,179 @@
+)
+
+var validUserPrivileges = NewPrivilegeSet(
+	privilegeMonitor,
+	privilegeOwnership,
+)
+var userGrantSchema = map[string]*schema.Schema{
+	"user_name": {
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "The name of the user on which to grant privileges.",
+		ForceNew:    true,
+	},
+	"privilege": {
+		Type:         schema.TypeString,
+		Optional:     true,
+		Description:  "The privilege to grant on the user.",
+		Default:      privilegeUsage.String(),
+		ForceNew:     true,
+		ValidateFunc: validation.StringInSlice(validUserPrivileges.ToList(), true),
+	},
+	"roles": {
+		Type:        schema.TypeSet,
+		Elem:        &schema.Schema{Type: schema.TypeString},
+		Optional:    true,
+		Description: "Grants privilege to these roles.",
+	},
+	"with_grant_option": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "When this is set to true, allows the recipient role to grant the privileges to other roles.",
+		Default:     false,
+		ForceNew:    true,
+	},
+	"enable_multiple_grants": {
+		Type:        schema.TypeBool,
+		Optional:    true,
+		Description: "When this is set to true, multiple grants of the same type can be created. This will cause Terraform to not revoke grants applied to roles and objects outside Terraform.",
+		Default:     false,
+	},
+}
+
+// UserGrant returns a pointer to the resource representing a user grant
+func UserGrant() *TerraformGrantResource {
+	return &TerraformGrantResource{
+		Resource: &schema.Resource{
+			Create: CreateUserGrant,
+			Read:   ReadUserGrant,
+			Delete: DeleteUserGrant,
+			Update: UpdateUserGrant,
+
+			Schema: userGrantSchema,
+			// FIXME - tests for this don't currently work
+			Importer: &schema.ResourceImporter{
+				StateContext: schema.ImportStatePassthroughContext,
+			},
+		},
+		ValidPrivs: validUserPrivileges,
+	}
+}
+
+// CreateUserGrant implements schema.CreateFunc
+func CreateUserGrant(d *schema.ResourceData, meta interface{}) error {
+	w := d.Get("user_name").(string)
+	priv := d.Get("privilege").(string)
+	grantOption := d.Get("with_grant_option").(bool)
+	builder := snowflake.UserGrant(w)
+	roles := expandStringList(d.Get("roles").(*schema.Set).List())
+
+	err := createGenericGrant(d, meta, builder)
+	if err != nil {
+		return err
+	}
+
+	grant := &grantID{
+		ResourceName: w,
+		Privilege:    priv,
+		GrantOption:  grantOption,
+		Roles:        roles,
+	}
+	dataIDInput, err := grant.String()
+	if err != nil {
+		return err
+	}
+	d.SetId(dataIDInput)
+
+	return ReadUserGrant(d, meta)
+}
+
+// ReadUserGrant implements schema.ReadFunc
+func ReadUserGrant(d *schema.ResourceData, meta interface{}) error {
+	grantID, err := grantIDFromString(d.Id())
+	if err != nil {
+		return err
+	}
+	w := grantID.ResourceName
+	priv := grantID.Privilege
+
+	err = d.Set("user_name", w)
+	if err != nil {
+		return err
+	}
+	err = d.Set("privilege", priv)
+	if err != nil {
+		return err
+	}
+	err = d.Set("with_grant_option", grantID.GrantOption)
```

Comment:
is it possible to read the role to determine if with_grant_option has been set already? it would be nice if this could actually be read, rather than just set from the id. i know we don't do that anywhere else, i am just wondering.

Reply:
propably yes, i beleive it's better to make in separte PR for it to have the same approach for all grants resources